### PR TITLE
Openedx nodebb discussion django app

### PR DESCRIPTION
**Pull request by mumarkhan999**
Contains the initial code to add a new `course tab` in openedx. And also shows `nodebb` in an `iframe`
which is visible when we click on `nodebb discussion tab`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+---
+
+2.  Copy this folder and place this at the following location in edx codebase
+
+    `../edx-platform/openedx/features/`
+
+---
+
+3.  Add `ENABLE_NODEBB_DISCUSSION` in the following files
```
Comment (asamolion) [inline README.md]: This flag should be added under the `FEATURES` key in the settings file

Comment (asamolion) [reply README.md]: Change the relevant code as well.

File: __init__.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+default_app_config = 'openedx.features.openedx_nodebb_discussion.apps.OpenedxNodebbDiscussionConfig'
+
+from django.conf import settings
+
+
+def is_feature_enabled():
```
Comment (asamolion) [inline __init__.py]: Why is this necessary?

File: middleware.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    Used for session sharing with NodeBB community
+    """
+
+    def process_response(self, request, response):
+        try:
+            if request.user.is_authenticated():
+                encoded_jwt = jwt.encode({'id': request.user.id,
+                                          'username': request.user.username,
+                                          'email': request.user.email},
+                                         'secret', algorithm='HS256')
```
Comment (asamolion) [inline middleware.py]: JWT secret and algorithm should be taken from env config.

File: middleware.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+                response.set_cookie('token', encoded_jwt, domain="localhost")
+            else:
+                response.delete_cookie('token', domain="localhost")
```
Comment (asamolion) [inline middleware.py]: domain should be taken from env as well.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    }
+```
+
+---
+
+7.  After that change the `version number` which is availabe in the same file for example if it is 
+    `0.11` change it to `0.12` and then go to `lms-shell` and your directory path should 
+    be like this
+
+    `/edx/app/edxapp/edx-platform#`
```
Comment (danialmalik) [inline README.md]: Why this whole directory hierarchy is necessary?

File: apps.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+class OpenedxNodebbDiscussionConfig(AppConfig):
+    """
+        Application Configuration for OPENEDX_NODEBB_DISCUSSION.
+    """
+    name = 'openedx.features.openedx_nodebb_discussion'
+
+    def ready(self):
+        from django.conf import settings
+        settings.MIDDLEWARE_CLASSES.append(
+            'openedx.features.openedx_nodebb_discussion.middleware.UserSessionSharingMiddleware')
```
Comment (danialmalik) [inline apps.py]: Please move the `)` to the next line (PEP8).

File: plugins.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    type = "openedx_nodebb_discussion"
+    title = ugettext_noop("NodeBB Discussion")
+    view_name = "nodebb_dashboard"
+
+    @classmethod
+    def is_enabled(cls, course, user=None):
+        if not super(NodeBBTab, cls).is_enabled(course, user):
+            return False
+
+        return settings.FEATURES.get('ENABLE_NODEBB_DISCUSSION', False)
```
Comment (danialmalik) [inline plugins.py]: Shouldn't we directly use this line only? The condition above is doing the same thing I guess.

File: views.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+from django_comment_client.utils import has_discussion_privileges
+from opaque_keys.edx.keys import CourseKey
+from student.models import CourseEnrollment
+
+from . import is_feature_enabled
+
+
+class NodebbDashboardView(TemplateView):
+    """
+    View methods related to the teams dashboard.
```
Comment (danialmalik) [inline views.py]: Please correct the docstring.

File: README.md (hunk 5 above)
Comment (imhassantariq) [reply README.md]: Because we are creating a new tab in our edx for courses and it is necessary to register tab in setup.py. In setup.py we have only 3 main directories lms,cms,openedx so we have to place our app in these directories.

File: apps.py (hunk 6 above)
Comment (mumarkhan999) [reply apps.py]: done.

File: views.py (hunk 8 above)
Comment (mumarkhan999) [reply views.py]: done.

File: middleware.py (hunk 4 above)
Comment (imhassantariq) [reply middleware.py]: done

File: middleware.py (hunk 3 above)
Comment (imhassantariq) [reply middleware.py]: done

File: __init__.py (hunk 2 above)
Comment (imhassantariq) [reply __init__.py]: Not necessary, Removed.

File: README.md (hunk 1 above)
Comment (imhassantariq) [reply README.md]: Done

File: README.md (hunk 5 above)
Comment (imhassantariq) [reply README.md]: edX install in this hierarchy. Please see this [Reference](https://openedx.atlassian.net/wiki/spaces/AC/pages/30965919/Adding+a+new+course+tab)

File: views.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class NodebbDashboardView(TemplateView):
+    """
+    View methods related to the nodebb dashboard.
+    """
+
+    def get(self, request, course_id):
+        """
+        Renders the teams dashboard, which is shown on the "Teams" tab.
```
Comment (danialmalik) [inline views.py]: Please correct the doc string.

File: middleware.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    Used for session sharing with NodeBB community
+    """
+
+    def process_response(self, request, response):
+        try:
+            if request.user.is_authenticated():
+                encoded_jwt = jwt.encode({'id': request.user.id,
+                                          'username': request.user.username,
+                                          'email': request.user.email},
+                                         django_settings.OPENEDX_NODEBB_DISCUSSION['SECRET'],
```
Comment (asamolion) [inline middleware.py]: `SECRET` is not that descriptive of a name. Use something more descriptive like `DISCUSSION_JWT_SECRET`

File: views.py (hunk 9 above)
Comment (imhassantariq) [reply views.py]: Done

File: middleware.py (hunk 10 above)
Comment (imhassantariq) [reply middleware.py]: Done
